Refactor gtda/diagrams

ulupo opened the pull request:
**Reference issues/PRs**
Fixes #438 (amongst other things).

**Types of changes**
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

**Description**
- **[bug fix]** Extends fix of y-axis as in #453 to `PersistenceImage`.

- **[enhancement]** Represents possible multiplicity of persistence pairs in persistence diagram plots by appending ", multiplicity: i" to the hovertext for points with multiplicity i.

- **[breaking change]** `gaussian_filter` (used to make heat kernel representations and persistence image ones) is now called with `mode="constant"` instead of `"reflect"`.

- **[bug fix/enhancement]** The `PersistenceLandscape` plot method is fixed:

  - Only the figure corresponding to the first seen homology dimension was previously returned.

  - Now, the output is a figure with subplots, a main title, and one subtitle per plot (homology dimension).

- **[breaking change]** The default value of `order` in `Amplitude` has been changed from `2.` to `None` (giving vector instead of scalar features)

- **[breaking change]** The meaning of the default `None` for `weight_function` in `PersistenceImage` (and `Amplitude` and `PairwiseDistance` when `metric="persistence_image"`) has been changed from the identity function (now removed) to the function returning a vector of ones. Reason: `None` corresponding the identity meant that there was a non-trivial weighting. Semantically, this does not seem correct ("None" should mean no weighting at all).

- Important changes to `gtda.diagrams._metrics`:

  - **[breaking change/enhancement]** The computation of heat/persistence image distances and amplitudes has been changed to yield the continuum limit when `n_bins` tends to infinity.

  - **[breaking change/enhancement]** `sigma` in persistence-image-- and heat-kernel--based representations/distances/amplitudes is now measured in the same units as the filtration parameter (not in pixels), thus decoupling its effect from the effect of `n_bins`. Also the default value of this parameter in `PairwiseDistance`, `Amplitude`, `HeatKernel` and `PersistenceImage` has been changed from 1. to 0.1.

  - **[enhancement]** Trivial points are now removed from diagrams before creating a sampled image in `heats` and `persistence_images`. This ensures in particular that the trivial points really give no contribution regardless of the weight function used for persistence images.

  - **[bug fix]** #438 is fixed, as are similar issues with `PairwiseDistance` when the metric is persistence-image--based.

  - **[enhancement]** `silhouettes` now does not create NaNs when a subdiagram is trivial (this was due to division 0/0, now fixed by a cheap trick making the denominator infinity).

  - **[breaking change/bug fix]** The default value of `power` in `silhouette_amplitudes` and `silhouette_distances` is changed to 1., to agree with `Amplitude` and `PairwiseDistance` docs.

  - **[bug fix]** The use of `np.array_equal` in `persistence_image_distances` and `heat_distances` is fixed -- so far the boolean check for equal inputs always failed due to the in-place modifications of `diagrams_1`.

  - **[breaking change]** Weights are no longer stored in the `effective_metric_params_` attribute of `PairwiseDistance`, `Amplitude` and `Scaler` objects when the metric is persistence-image--based. The weight function is.

  - The utility function `_heat` is removed and its logic is incorporated in `heats`.
  
  - The utility functions `_matrix_wrapper` and `_arrays_wrapper` were never used and have been removed.

- **[breaking change?]** The `homology_dimensions_` attributes of several transformers have been converted from lists to tuples and, when possible, homology dimensions are cast to ints there. Also, the keys of the `samplings_` dictionary attributes of several transformers are now cast to ints when possible.

- **[enhancement]** Tests of plot methods in `gtda.diagrams.representations` have been improved to cover the use of `plotly_params` and of other kwargs, and the case of diagrams with homology dimension equaling `np.inf`.

- **[enhancement]** Test coverage of `Amplitude` and `PairwiseDistance` has been improved as follows:

  - Silhouettes and persistence images are covered throughout.

  - The `order` parameter is covered throughout.

- **[enhancement]** Code style and clarity have been improved throughout.

- **[enhancement]** Miscellaneous docstring and linting improvements have been performed.

**Screenshots (if appropriate)**

**Any other comments?**

**Checklist**
- [x] I have read the [guidelines for contributing](https://giotto-ai.github.io/gtda-docs/latest/contributing/#guidelines).
- [x] My code follows the code style of this project. I used `flake8` to check my Python changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed. I used `pytest` to check this on Python tests.

ulupo committed:
First attempt at Amplitude fix

ulupo committed:
First attempt at Amplitude fix

ulupo committed:
Merge remote-tracking branch 'origin/fix_amplitude' into fix_amplitude

ulupo committed:
Use output kwarg of scipy's gaussian_filter, be explicit about dtypes

ulupo committed:
Fix y-axis in PersistenceImage plots

ulupo committed:
Style improvements

ulupo committed:
Duplicate hovertext on persistence pairs with multiplicity in plot_diagram

ulupo committed:
Minor improvements with slicing, naming and integer conversions in gtda.diagrams._metrics

ulupo committed:
Change reflect mode of gaussian_filter to "constant"

Affects HeatKernel and PersistenceImage

ulupo committed:
Fix PersistenceLandscape plot method

- Only the figure corresponding to the first seen homology dimension was returned
- Output is now a figure with subplots, a main title, and one subtitle per plot (homology dimension)

ulupo committed:
Improve tests for plot methods in gtda.diagrams.representations

Cover use of plotly_params

ulupo committed:
Minor docstring linting

ulupo committed:
Miscellaneous docstring improvements in gtda/diagrams

ulupo committed:
Fix validation dictionary for metric_params in the case of PersistenceImage

ulupo committed:
Change default value of `order` in Amplitude, from 2. to None (vector features)

ulupo committed:
Change meaning of default None for weight_function in PersistenceImage

- ``None`` corresponding the identity means that there really is a non-trivial weighting in that case. Semantically, this does not seem correct ("None" should mean no weighting at all)

ulupo committed:
Improve code style and clarity in plot methods in gtda.diagrams.representations

ulupo committed:
Refactor gtda/diagrams/_metrics.py to fix several bugs

- Change computation of heat/persistence image distances and amplitudes to yield the continuum limit when `n_bins` tends to infinity.
- Make `sigma` in persistence-image-- and heat-kernel--based representations/distances/amplitudes measured in the same units as the filtration parameter (not in pixels), thus decoupling its effect from the effect of `n_bins`. Also change the default value in PairwiseDistance, Amplitude, HeatKernel and PersistenceImage from 1. to 0.1.
- Remove trivial points from diagrams before creating a sampled image in `heats` and `persistence_images`. This ensures in particular that the trivial points really give no contribution regardless of the weight function used for persistence images.
- Finish fixing #438 and similar issues with PairwiseDistance when the metric is persistence-image--based.
- Ensure `silhouettes` does not create NaNs when a subdiagram is trivial.
- Change default value of `power` in `silhouette_amplitudes`
 and `silhouette_distances` to 1., to agree with Amplitude and PairwiseDistance docs.
- Fix use of np.array_equal in `persistence_image_distances` and `heat_distances` -- so far the boolean check for equal inputs always failed due to the in-place modifications of `diagrams_1`.
- No longer store weights in `effective_metric_params_` attribute of PairwiseDistance and Amplitude when the metric is persistence-image--based.
- Remove _heat from gtda.diagrams._metrics.
- Remove identity from gtda.diagrams._utils and make default `weight_function` kwargs equal to np.ones_like everywhere to agree with default in `PersistenceImage`.
- Other style improvements (variable names, linting)

ulupo committed:
Fix trace name when homology dimension is np.inf in BettiCurve and Silhouette

ulupo committed:
Improve tests of plot methods in representations

Also adapt test_all_pts_the_same to new behaviour of `heats_` in gtda.diagrams._metrics

ulupo committed:
Improve test coverage of Amplitude and PairwiseDistance

- Make sure silhouettes and persistence images are covered throughout
- Cover `order` parameter throughout

wreise commented on gtda/diagrams/distance.py:
```diff
@@ -37,22 +34,29 @@ class PairwiseDistance(BaseEstimator, TransformerMixin):
+
+        - Input collections of persistence diagrams for this transformer must
+          satisfy certain requirements, see e.g. :meth:`fit`.
+        - The shape of outputs of :meth:`transform` depends on the value of the
```
Oh yes!

wreise commented on gtda/diagrams/representations.py:
```diff
@@ -450,59 +448,67 @@ def plot(self, Xt, sample=0, homology_dimensions=None, plotly_params=None):
+                          for _, dim in _homology_dimensions]
+        fig = make_subplots(rows=len(_homology_dimensions), cols=1,
+                            subplot_titles=subplot_titles)
+        many_homology_dim = len(_homology_dimensions) - 1
```
Maybe something that starts with `is_` or `has_` to indicate that it's used as a boolean?

ulupo replied:
Done.

wreise commented on gtda/diagrams/representations.py:
```diff
@@ -1156,44 +1171,43 @@ def plot(self, Xt, sample=0, homology_dimensions=None, plotly_params=None):
                     ix = np.flatnonzero(homology_dimensions_arr == dim)[0]
                     _homology_dimensions.append((ix, dim))
 
+        layout_axes_common = {
+            "type": "linear",
+            "ticks": "outside",
+            "showline": True,
+            "zeroline": True,
+            "linewidth": 1,
+            "linecolor": "black",
+            "mirror": False,
+            "showexponent": "all",
+            "exponentformat": "e"
+            }
```
Can this be extracted to a dictionary common across multiple transformers?

ulupo replied:
It could be placed somewhere in `gtda/plotting` perhaps.

ulupo replied:
... But I'd prefer to leave this to another PR (just to avoid extending beyond `diagrams` for now).

ulupo committed:
Improve variable name after @wreise's suggestion

ulupo committed:
Add test of zero weight_function for PersistenceImage

ulupo committed:
Make behaviour of `Scaler.fit` when the metric is persistence image the same as Amplitude

Accordingly add more combinations of metrics and metric_params in tests for Scaler

ulupo committed:
Delete never-used _matrix_wrapper and _arrays_wrapper functions

wreise commented on gtda/plotting/persistence_diagrams.py:
```diff
@@ -63,8 +63,18 @@ def plot_diagram(diagram, homology_dimensions=None, plotly_params=None):
+            subdiagram, axis=0, return_inverse=True, return_counts=True
+            )
+        hovertext = [
+            "<br>".join([f"{tuple(unique[unique_row_index][:2])}"]
```
I like the idea of including multiplicity. However, I am worried that this might be unreadable when the multiplicity is >10 (this can be the case for images).

wreise replied:
What about adding it in the however text if the multiplicity is greater than 1?
For example:
"(2,3)" or "(2,3), multiplicity: 5",

ulupo replied:
A great suggestion!

ulupo replied:
Done.

ulupo committed:
Remove _pad from gtda.diagrams._utils as it is never used

ulupo committed:
Make `copy=True` in calls to check_diagrams in Scaler.transform and Scaler.inverse_transform

wreise commented on gtda/diagrams/_metrics.py:
```diff
@@ -74,55 +82,76 @@ def landscapes(diagrams, sampling, n_layers):
+
+    # diagrams_ax[...] = (diagrams_ax - first_sampling) / step_size[ax]
+    #
+    # diagrams = diagrams.astype(int)
```
Should these be here?

ulupo replied:
You're right! Fixed.

wreise commented on gtda/diagrams/_metrics.py:
```diff
@@ -74,55 +82,76 @@ def landscapes(diagrams, sampling, n_layers):
+        image = persistence_images_[i]
+        _sample_image(image, diagram_nontrivial_pixel_coords)
+        image *= weights
+        gaussian_filter(image, sigma_pixel, mode="constant", output=image)
```
Seems like the smoothing is applied twice? (lines 150-151 a well)

ulupo replied:
Oh dear! Well spotted!!

ulupo replied:
Should be fixed now.

ulupo committed:
Make homology_dimensions_ attributes tuples instead of lists, with integers when possible

ulupo committed:
Remove forgotten commented lines

ulupo committed:
Avoid applying smoothing twice in persistence_images following @wreise's review

ulupo committed:
Implement suggestion by @wreise to avoid excessive hovertext in plot_diagrams

See https://github.com/giotto-ai/giotto-tda/pull/454#discussion_r469084946

ulupo committed:
Fix small bug introduced in Filtering in 17b906e3d5677b9c28a5c782783b16d5a5d448db, add new tests for Filtering

ulupo committed:
Improve code style

ulupo committed:
Hard-code zero array outputs by `heats` and `persistence_images` when step sizes are zero

Modify test_all_pts_the_same accordingly

ulupo committed:
Add `homology_dimensions` kwarg to `_bin`

Achieves beautification of self._samplings and self.samplings_ (ints shown instead of floats) in several transformers (and saves some computation)

ulupo committed:
Adapt choices of min_values, max_values and sigmas in hypothesis-based tests

New meaning of sigma led to overflow issues in existing tests.

ulupo committed:
Fix linting

ulupo committed:
Minor style improvements

ulupo committed:
Add useful inline comments

ulupo committed:
Make tests of HeatKernel and PersistenceImage less flaky

ulupo committed:
Typo fix

ulupo committed:
Make all homology dimensions equal in test_hk_big_sigma

Also extend this test to PersistenceImage, and rename it accordingly

ulupo committed:
Cover use of `plotly_params` kwarg in diagram preprocessing classes plot methods

ulupo committed:
Simplify plot code following 17b906e3d5677b9c28a5c782783b16d5a5d448db

ulupo committed:
Extract some common logic from plot methods in gtda.diagrams.representations

ulupo committed:
Fix typo

ulupo committed:
Fix linting

ulupo committed:
Silence expected warnings from image transformers in test_common

wreise commented on gtda/diagrams/_metrics.py:
```diff
@@ -284,34 +343,60 @@ def wasserstein_amplitudes(diagrams, p=2., **kwargs):
+    step_sizes_factor = np.product(step_size) ** (1 / p)
+    # WARNING: `persistence_images` modifies `diagrams` in place
+    persistence_images_ = persistence_images(
+        diagrams, sampling, step_size, sigma, weights
+        ).reshape(len(diagrams), -1)
```
I like how you consistently reshape into a 2d array - super clear where the norm is then.

ulupo replied:
Right. Not only that, the current approach was also incorrect! It was computing the 2-norm of matrices, which is the absolute value of the largest singular value. But we wanted the Frobenius norm in that case...

ulupo replied:
And it failed when p was a float for that reason...

wreise commented on gtda/diagrams/_utils.py:
```diff
@@ -10,8 +10,10 @@ def _subdiagrams(X, homology_dimensions, remove_dim=False):
     the same number of points in each homology dimension."""
     n = len(X)
     if len(homology_dimensions) == 1:
+        # Reshape ensures copy
```
Let me present what i found.
Reading the documentation (see [the Returns section](https://numpy.org/doc/stable/reference/generated/numpy.reshape.html)), it seems that it returns a copy only if it cannot reshape the diagram. So, for example, if `all(X[:, :, 2] == homology_dimensions[0])`, then the shape would not be changed and the array might not be copied.

ulupo replied:
I tested precisely your scenario and found that this was not the case. The reason is that `X[X[:, :, 2] == dim]` can never be of the same shape as `X. Could you double check?

ulupo replied:
Let me also say that it is *essential* that `_subdiagrams` always returns copies, as we do all sorts of in-place things to subdiagrams in several transformers.

wreise replied:
True, my bad! I forgot what the output of `X[X[:, :, 2] == dim]` looks like.

wreise commented on gtda/diagrams/_utils.py:
```diff
@@ -131,18 +127,30 @@ def _bin(X, metric, n_bins=100, **kw_args):
-    weights = {dim: weight_function(samplings[dim][:, 1])
-               for dim in samplings.keys()}
-    return weights
+def _make_homology_dimensions_mapping(homology_dimensions,
+                                      homology_dimensions_ref):
+    """`homology_dimensions_ref` is assumed to be a sorted tuple as is e.g.
+    :attr:`homology_dimensions_` for several transformers."""
+    if homology_dimensions is None:
+        homology_dimensions_mapping = list(enumerate(homology_dimensions_ref))
+    else:
+        homology_dimensions_mapping = []
+        for dim in homology_dimensions:
+            if dim not in homology_dimensions_ref:
+                raise ValueError(f"All homology dimensions must be in "
+                                 f"{homology_dimensions_ref}; {dim} is not.")
+            else:
+                homology_dimensions_arr = np.array(homology_dimensions_ref)
+                inv_idx = np.flatnonzero(homology_dimensions_arr == dim)[0]
+                homology_dimensions_mapping.append((inv_idx, dim))
+    return homology_dimensions_mapping
```
I am not sure I understand what this does exactly. To ,y understanding, we map the homology dimensions provided by the user to the indices of those dimension in the transformer. So, if the trasnformer has `homology_dimensions_ref=[0,1,4, 6]` and the user passes `homology_dimension=[4]` to plot, the output will be `[(2,4)]`?

ulupo replied:
That's right. This piece of code was there before, I just moved it to `_utils` to avoid repeating it twice.

wreise commented on gtda/diagrams/features.py:
```diff
@@ -126,7 +127,10 @@ def fit(self, X, y=None):
+        self.homology_dimensions_ = tuple(
+            sorted([int(dim) if dim != np.inf else dim
+                    for dim in set(X[0, :, 2])])
+            )
```
This logic appears a total of 7 times in the code. It's pretty simple but I what would you say about extracting it to a common function `homology_dimensions_to_ints`, in `gtda.diagrams._utils`?

ulupo replied:
Yeah, that could be done!

ulupo replied:
Implemented in https://github.com/giotto-ai/giotto-tda/pull/454/commits/c2367aa3420702b08421f99285555a0307c4f657, could you have a look?

wreise replied:
Yes!
Oh, I just realized :facepalm: that we look for dimensions only in the first diagram! What about something like `set(np.ravel(X[:,:,2]))`?

ulupo replied:
This is by design, as my new inline comments and the docstrings explain. I think it's correct like this, unless I'm missing something.

ulupo replied:
I am about to suggest a change to `_subdiagrams` to be even more informative to the user about this.

wreise replied:
Yes - good point!

wreise commented on gtda/diagrams/features.py:
```diff
@@ -322,15 +330,21 @@ def fit(self, X, y=None):
+                'weight_function', None
+                )
+            if weight_function is None:
+                self.effective_metric_params_['weight_function'] = np.ones_like
```
I missed something: if `weight_function` is not `None`, we don't use it? ( it's not passed to self.effective_metric_params_['weight_function'])

ulupo replied:
You're right, excellent! This is the result of a bad copy-paste. I'll fix this and add a test to make sure it never happens again (I'll test the zero weight-function to check that it yields zero amplitudes and pairwise distances).

ulupo replied:
Done in 4001abf

wreise commented on gtda/diagrams/preprocessing.py:
```diff
@@ -237,15 +239,21 @@ def fit(self, X, y=None):
+                'weight_function', None
+                )
+            if weight_function is None:
+                self.effective_metric_params_['weight_function'] = np.ones_like
```
Same here.

ulupo replied:
Thanks. And in `PairwiseDistance` too I believe.

ulupo replied:
Done in 4001abf

wreise reviewed (commented):
Huge, and quite a few nice changes!

ulupo committed:
Implement @wreise's suggestion to abstract away sorting and integer conversion of fit hom dims

ulupo committed:
Linting

ulupo committed:
Reintroduced accidentally deleted line in Silhouette

ulupo committed:
Fix use of non-default weight functions in Amplitude, PairwiseDistance and Filtering with persistence-image metric

Also add tests for Amplitude and PairwiseDistance to check that a zero weight function yields identically zero amplitudes/distances.

ulupo committed:
Refactor `_subdiagrams` to be able to throw informative errors on expected input properties

ulupo committed:
Fix variable name

wreise commented on gtda/diagrams/distance.py:
```diff
@@ -192,8 +192,9 @@ def fit(self, X, y=None):
-            if weight_function is None:
-                self.effective_metric_params_['weight_function'] = np.ones_like
+            weight_function = \
+                np.ones_like if weight_function is None else weight_function
```
Isn't this equivalent to `weight_function = self.effective_metric_params_.get( 'weight_function', np.ones_like)`?

wreise replied:
Or, `self.effective_metric_params_.setdefault(np.ones_like)`?

ulupo replied:
Not quite I think, because I think we'd like to allow the user to place `None` and yet we have to pass `np.ones_like` to `_parallel_amplitude` etc.

ulupo replied:
But maybe I'm wrong again about this code section :)

wreise replied:
you're right!

wreise reviewed (approved):
Ok, it's great for me!
